api/login.go

```diff
@@ -0,0 +1,440 @@
+		// The result contains a discharge-required
+		// macaroon. We discharge it and retry
+		// the login request with the original macaroon
+		// and its discharges.
+		if result.DischargeRequiredReason == "" {
+			result.DischargeRequiredReason = "no reason given for discharge requirement"
+		}
+		// Prefer the newer bakery.v2 macaroon.
+		dcMac := result.BakeryDischargeRequired
+		if dcMac == nil {
+			dcMac, err = bakery.NewLegacyMacaroon(result.DischargeRequired)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+		if err := p.bakeryClient.HandleError(ctx, p.cookieURL, &httpbakery.Error{
+			Message: result.DischargeRequiredReason,
+			Code:    httpbakery.ErrDischargeRequired,
+			Info: &httpbakery.ErrorInfo{
+				Macaroon:     dcMac,
+				MacaroonPath: "/",
+			},
+		}); err != nil {
+			cause := errors.Cause(err)
+			if httpbakery.IsInteractionError(cause) {
+				// Just inform the user of the reason for the
+				// failure, e.g. because the username/password
+				// they presented was invalid.
+				err = cause.(*httpbakery.InteractionError).Reason
+			}
+			return nil, errors.Trace(err)
+		}
+		// Add the macaroons that have been saved by HandleError to our login request.
+		request.Macaroons = httpbakery.MacaroonsForURL(p.bakeryClient.Jar, p.cookieURL)
+		result = params.LoginResult{} // zero result
+		err = caller.APICall("Admin", 3, "", "Login", request, &result)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if result.DischargeRequired != nil {
+			return nil, errors.Errorf("login with discharged macaroons failed: %s", result.DischargeRequiredReason)
+		}
+	}
+
+	var controllerAccess string
+	var modelAccess string
+	tag := p.tag
+	if result.UserInfo != nil {
+		tag, err = names.ParseTag(result.UserInfo.Identity)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		controllerAccess = result.UserInfo.ControllerAccess
+		modelAccess = result.UserInfo.ModelAccess
+	}
+	servers := params.ToMachineHostsPorts(result.Servers)
+	serverVersion, err := version.Parse(result.ServerVersion)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return &LoginResultParams{
+		tag:              tag,
+		modelTag:         result.ModelTag,
+		controllerTag:    result.ControllerTag,
+		servers:          servers,
+		publicDNSName:    result.PublicDNSName,
+		facades:          result.Facades,
+		modelAccess:      modelAccess,
+		controllerAccess: controllerAccess,
+		serverVersion:    serverVersion,
+	}, nil
+}
+
+// NewAnyOfLoginProvider returns a login provider that will attempt to
+// log in using all the specified login providers in sequence - results
+// of the first will be returned.
+// This login provider should only be used when connecting to a controller
+// for the first time when we still don't know which login method.
+func NewAnyOfLoginProvider(providers ...LoginProvider) LoginProvider {
+	return &anyOfLoginProviders{
+		providers: providers,
+	}
+}
+
+type anyOfLoginProviders struct {
```

Comment:
Can we call this `orderedLoginProviders` or `tryInOrderLoginProviders` or `prioritisedLoginProviders` or something to make it clear order matters. `Any` implies try them in some undetermined order and whatever gets there is used.

Reply:
Ideally we'd probe an unauthenticated login endpoint to get an ordered list of supported login types. Have we thought about that approach? And if we get back a "not implemented" error we fall back to UserPass.

Reply:
we are doing something similar with this approach.. trying oauth2 first, if not implemented falling back to user-pass.. would a separate endpoint have any value? are we planning on supporting more login providers?

Reply:
We are doing it ordered as you say, but the name of the struct does not reflect that and implies to the casual observed that order does not matter.
As discussed elsewhere, allowing the client probe the auth server for what is supported is a cleaner approach that we should adopt. That way, the client does not waste time and resources trying oauth2 when the server does not even support it, eg older controllers. We can make progress by landing this but we should not go to release without the login probe being done.